File: src/offer24.ReverseList/reverse_list.cpp
/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution1 {
public:
    ListNode* ReverseList(ListNode* pHead) {
        ListNode* pPrev = nullptr;
        ListNode* pNode = pHead;
        ListNode* pReversedHead = nullptr;
        while (pNode != nullptr){
            ListNode* pNext = pNode->next;
            if (pNext == nullptr)
                pReversedHead = pNode;
            pNode->next = pPrev;
            pPrev = pNode;
            pNode = pNext;
        }
        return pReversedHead;
    }
};

/*
struct ListNode {
	int val;
	struct ListNode *next;
	ListNode(int x) :
			val(x), next(NULL) {
	}
};*/
class Solution2 {
public:
    ListNode* ReverseList(ListNode* pHead) {
        if (pHead->next == nullptr || pHead == nullptr){
            return pHead;
        }
        ListNode* pReversedNode = ReverseList(pHead->next);
        pHead->next->next = pHead;
        pHead->next = nullptr;
        return pReversedNode;
    }
};

File: src/offer24.ReverseList/reverse_list.java
//1
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode ReverseList(ListNode head) {
        ListNode pre = null;
        ListNode nxt = head;
        ListNode cur = head;
        while(nxt != null){
            nxt = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
}
//2
/*
public class ListNode {
    int val;
    ListNode next = null;

    ListNode(int val) {
        this.val = val;
    }
}*/
public class Solution {
    public ListNode ReverseList(ListNode head) {
        if(head == null || head.next == null){
            return head;
        }
        ListNode node = ReverseList(head.next);
        head.next.next = head;
        head.next = null;
        return node;
    }
}